File: mathematics/matrix/csharp/mathematics/Matrix.cs
using System.Collections.Generic;
using System.Numerics;

namespace mathematics;

public sealed class Matrix <T>  where T: INumber<T>
{
    public int ColumnsCount{get; set;}
    public int RowsCount{get; set;}

    private List<List<T>> matrix_;


    public Matrix(int rowsCount, int columnsCount){
        RowsCount = rowsCount;
        ColumnsCount = columnsCount;
        List<List<T>> newMatrix = new List<List<T>>();
        for (int i = 0; i < rowsCount; i++){
            newMatrix.Add(Enumerable.Repeat(default(T), columnsCount).ToList()!);
        }
        matrix_ = newMatrix;
    }

    public Matrix(List<List<T>> matrix){
        RowsCount = matrix.Count;
        ColumnsCount = matrix[0].Count;
        matrix_ = matrix;
    }

    public Matrix<T> Dot(Matrix<T> m){
        if(ColumnsCount != m.RowsCount)
            throw new Exception("Matrix size columns not equil size rows!");
        List<List<T>> dot_matrix = new List<List<T>>();

        for (int i = 0; i < RowsCount; i++){
            List<T> n_row = new List<T>();
            for (int j = 0; j < m.ColumnsCount; j++){
                T? sum = default!;
                for (int k = 0; k < m.RowsCount; k++){
                    sum += matrix_[i][k] * m.matrix_[k][j];
                }
                n_row.Add(sum);
            }
            dot_matrix.Add(n_row);
        }

        Matrix<T> res = new Matrix<T>(dot_matrix);
        return res;
    }


    public static Matrix<T> operator * (Matrix<T> m1, Matrix<T> m2){
        Matrix<T> res = m1.Dot(m2);
        return res;
    }


    public static Matrix<T> operator * (Matrix<T> m, dynamic n){
        Matrix<T> res = new Matrix<T>(m.RowsCount, m.ColumnsCount);
        for(int i = 0; i < m.RowsCount; i++){
            for(int j = 0; j < m.ColumnsCount; j++){
                res.matrix_[i][j] = m.matrix_[i][j] * n;
            }
        }
        return res;
    }


    public static Matrix<T> operator - (Matrix<T> m1, Matrix<T> m2){

        if(m1.ColumnsCount != m2.ColumnsCount || m1.RowsCount != m2.RowsCount)
            throw new Exception("Matrix size columns not equil size rows!");

        Matrix<T> res = new Matrix<T>(m1.RowsCount, m1.ColumnsCount);
        for (int i = 0; i < m1.RowsCount; i++){
            for (int j = 0; j < m1.ColumnsCount; j++){
                res.matrix_[i][j] = m1.matrix_[i][j] - m2.matrix_[i][j];
            }
        }

        return res;
    }


    public static Matrix<T> operator + (Matrix<T> m1, Matrix<T> m2){

        if(m1.ColumnsCount != m2.ColumnsCount || m1.RowsCount != m2.RowsCount)
            throw new Exception("Matrix size columns not equil size rows!");

        Matrix<T> res = new Matrix<T>(m1.RowsCount, m1.ColumnsCount);
        for (int i = 0; i < m1.RowsCount; i++){
            for (int j = 0; j < m1.ColumnsCount; j++){
                res.matrix_[i][j] = m1.matrix_[i][j] + m2.matrix_[i][j];
            }
        }

        return res;
    }

    public void ShowMatrix(){
        for(int i = 0; i < RowsCount; i++){
            for(int j = 0; j < ColumnsCount; j++){
                Console.Write($"{matrix_[i][j]} \t");
            }
                Console.Write("\n");
        }
    }


    public Matrix<T> Transporate(){
        List<List<T>> transporateMatrix = new List<List<T>>();

        for(int i = 0; i < ColumnsCount; i++){
            List<T> t_column = new List<T>();
            for(int j = 0; j < RowsCount; j++){
                t_column.Add(matrix_[j][i]);
            }
            transporateMatrix.Add(t_column);
        }

        Matrix<T> res = new Matrix<T>(transporateMatrix);
        return res;
    }


}

File: mathematics/matrix/go/matrix/matrix.go
package matrix

import (
	"errors"
	"fmt"
)

type Number interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64 | ~float32 | ~float64 |
		~uint | ~uint32 | ~uint16 | ~uint64
}

type Matrix[T Number] struct {
	RowsCount    int
	ColumnsCount int
	Arr2D        [][]T
}

func (m *Matrix[T]) ShowMatrix() {
	for i := 0; i < m.RowsCount; i++ {
		for j := 0; j < m.ColumnsCount; j++ {
			fmt.Print(m.Arr2D[i][j], "\t")
		}
		fmt.Println()
	}
}

func (m *Matrix[T]) Add(m2 *Matrix[T]) (*Matrix[T], error) {
	if m.ColumnsCount != m2.ColumnsCount || m.RowsCount != m2.RowsCount {
		return nil, errors.New("matrix size columns not equil size rows")
	}

	mf := MatrixFabric[T]{}
	sum_m := mf.Create(m.RowsCount, m.ColumnsCount)

	for i := 0; i < m.RowsCount; i++ {
		for j := 0; j < m.ColumnsCount; j++ {
			sum_m.Arr2D[i][j] = m.Arr2D[i][j] + m2.Arr2D[i][j]
		}
	}

	return sum_m, nil
}

func (m *Matrix[T]) Sub(m2 *Matrix[T]) (*Matrix[T], error) {
	if m.ColumnsCount != m2.ColumnsCount || m.RowsCount != m2.RowsCount {
		return nil, errors.New("matrix size columns not equil size rows")
	}

	mf := MatrixFabric[T]{}
	sub_m := mf.Create(m.RowsCount, m.ColumnsCount)

	for i := 0; i < m.RowsCount; i++ {
		for j := 0; j < m.ColumnsCount; j++ {
			sub_m.Arr2D[i][j] = m.Arr2D[i][j] - m2.Arr2D[i][j]
		}
	}

	return sub_m, nil
}

func (m *Matrix[T]) Dot(m2 *Matrix[T]) (*Matrix[T], error) {
	if m.ColumnsCount != m2.RowsCount || m.RowsCount != m2.RowsCount {
		return nil, errors.New("matrix size columns not equil size rows")
	}

	dot_matrix := make([][]T, 0)

	for i := 0; i < m.RowsCount; i++ {
		n_row := make([]T, 0)
		for j := 0; j < m.ColumnsCount; j++ {
			var sum T
			for k := 0; k < m.RowsCount; k++ {
				sum += m.Arr2D[i][k] * m2.Arr2D[k][j]
			}
			n_row = append(n_row, sum)
		}
		dot_matrix = append(dot_matrix, n_row)
	}

	mf := MatrixFabric[T]{}
	res, err := mf.CreateByArr2D(dot_matrix)
	return res, err
}

func (m *Matrix[T]) Mul(scalar T) *Matrix[T] {
	mf := MatrixFabric[T]{}
	sub_m := mf.Create(m.RowsCount, m.ColumnsCount)

	for i := 0; i < m.RowsCount; i++ {
		for j := 0; j < m.ColumnsCount; j++ {
			sub_m.Arr2D[i][j] = m.Arr2D[i][j] * scalar
		}
	}

	return sub_m
}

func (m *Matrix[T]) Transporate() (*Matrix[T], error) {
	transporateMatrix := make([][]T, 0)

	for i := 0; i < m.ColumnsCount; i++ {
		t_column := make([]T, 0)
		for j := 0; j < m.RowsCount; j++ {
			t_column = append(t_column, m.Arr2D[j][i])
		}
		transporateMatrix = append(transporateMatrix, t_column)
	}

	mf := MatrixFabric[T]{}
	res, err := mf.CreateByArr2D(transporateMatrix)
	return res, err
}
